countItems.js

```diff
@@ -1,72 +1,165 @@
+            });
+            return callback(err);
+        }
+        return callback();
+    });
+}
 
-if (process.env.MONGODB_AUTH_USERNAME &&
-    process.env.MONGODB_AUTH_PASSWORD) {
-    params.authCredentials = {
-        username: process.env.MONGODB_AUTH_USERNAME,
-        password: process.env.MONGODB_AUTH_PASSWORD,
+function work(bucketList, callback) {
+    const { bucketInfos, bucketCount } = bucketList;
+    const retBucketInfos = bucketInfos.map(bucket => ({
+        name: bucket.getName(),
+        location: bucket.getLocationConstraint(),
+        isVersioned: !!bucket.getVersioningConfiguration(),
+        ownerCanonicalId: bucket.getOwner(),
+        ingestion: bucket.isIngestionBucket(),
+    }));
+    const store = {
+        objects: 0,
+        versions: 0,
+        buckets: bucketCount,
+        bucketList: retBucketInfos,
+        dataManaged: {
+            total: { curr: 0, prev: 0 },
+            byLocation: {},
+        },
+        stalled: 0,
     };
+    async.forEach(
+        bucketInfos,
+        (bucketInfo, done) =>
+            workers.countItems(bucketInfo, (err, results) => {
+                if (err) {
+                    return done(err);
+                }
+                if (results) {
+                    store.objects += results.objects;
+                    store.versions += results.versions;
+                    store.stalled += results.stalled;
+                    mongoClient.consolidateData(store, results.dataManaged);
+                }
+                return done();
+            }),
+        err => {
+            if (err) {
+                return callback(err);
+            }
+            return callback(null, store);
+        });
 }
 
-const mongoclient = new MongoClientInterface(params);
+function cleanup(store, callback) {
+    async.times(numWorkers, (_, done) => workers.clientTeardown(done), err => {
+        if (err) {
+            log.error('unable to tear down worker clients', {
+                error: err,
+            });
+        }
+        log.info('terminating workers');
+        workerFarm.end(workers);
+        if (callback) {
+            callback(err, store);
+        }
+    });
+}
 
-mongoclient.setup(err => {
-    if (err) {
-        log.error('error connecting to mongodb', {
-            error: err,
-        });
-        return process.exit(1);
-    }
-    return mongoclient.scanItemCount(log, error => {
-        // close connection before exiting
-        mongoclient.close();
-        if (error) {
-            log.error('error occurred during mongo scan item count', {
-                error,
+
+function scanItems(bucketList, callback) {
+    async.waterfall([
+        next => setup(next),
+        next => work(bucketList, next),
+    ], (err, store) => {
+        if (err) {
+            log.error('error occurred in worker process', {
+                error: err,
             });
-            return process.exit(1);
+            return cleanup(null, callback);
```

Comment:
The error seems to be lost here, won't this code path end up trying to call save with `null` results?

Reply:
you are completely correct. error is lost in the callback and `null` will be written